tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -82,241 +39,142 @@
-              + dtype
-              + ", got "
-              + dataTypeOf(obj)
-              + ")");
-    }
-    long[] dimSizes = new long[numDimensions(obj, dtype)];
-    fillShape(obj, 0, dimSizes);
-    Tensor<T> t = new Tensor<>(dtype, Shape.of(dimSizes));
-    TF_Tensor nativeHandle;
-    if (t.dtype != TString.DTYPE) {
-      long byteSize = elemByteSize(t.dtype) * t.shape.size();
-      nativeHandle = allocate(t.dtype.nativeCode(), dimSizes, byteSize);
-      setValue(nativeHandle, obj);
-    } else if (t.shape.numDimensions() != 0) {
-      nativeHandle = allocateNonScalarBytes(dimSizes, (Object[]) obj);
-    } else {
-      nativeHandle = allocateScalarBytes((byte[]) obj);
-    }
-    t.nativeRef = new NativeReference(nativeHandle);
-    return t;
+  public static <T extends TType> Tensor<T> of(DataType<T> dtype, Shape shape) {
+    return of(dtype, shape, shape.size() * dtype.byteSize());
   }
 
   /**
-   * Creates a tensor from an object whose class is inspected to figure out what the underlying data
-   * type should be.
+   * Allocates a tensor of a given datatype, shape and size.
    *
-   * @throws IllegalArgumentException if {@code obj} is not compatible with the TensorFlow type
-   *     system.
-   */
-  public static Tensor<?> create(Object obj) {
-    return create(obj, dataTypeOf(obj));
-  }
-
-  /**
-   * Create a {@link Integer} Tensor with data from the given buffer.
+   * <p>This method is identical to {@link #of(DataType, Shape)}, except that the final size of the
+   * tensor is explicitly set instead of computing it from the datatype and shape.
    *
-   * <p>Creates a Tensor with the given shape by copying elements from the buffer (starting from its
-   * current position) into the tensor. For example, if {@code shape = {2,3} } (which represents a
-   * 2x3 matrix) then the buffer must have 6 elements remaining, which will be consumed by this
-   * method.
+   * <p>This could be useful for tensor types that stores data but also metadata in the tensor memory,
+   * like {@link org.tensorflow.types.TString TString}.
    *
-   * @param shape the tensor shape.
-   * @param data a buffer containing the tensor data.
-   * @throws IllegalArgumentException If the tensor shape is not compatible with the buffer
+   * @param <T> the tensor element type
+   * @param dtype datatype of the tensor
+   * @param shape shape of the tensor
+   * @param size size, in bytes, of the tensor
+   * @return an allocated but uninitialized tensor
+   * @see #of(DataType, Shape)
    */
-  public static Tensor<TInt32> create(long[] shape, IntBuffer data) {
-    Tensor<TInt32> t = allocateForBuffer(TInt32.DTYPE, shape, data.remaining());
-    t.buffer().asIntBuffer().put(data);
+  public static <T extends TType> Tensor<T> of(DataType<T> dtype, Shape shape, long size) {
+    Tensor<T> t = new Tensor<>(dtype, shape);
+    TF_Tensor nativeHandle = allocate(t.dtype.nativeCode(), shape.asArray(), size);
```

Comment:
Does this call check that the size is larger than the required size for the shape & dtype?

Reply:
I think the native part validate it but I can add a unit test to confirm this

Reply:
Ok, so after some investigation, I can see that the native call returns null if the buffer is **smaller** than the required size (a larger buffer is accepted though so we should preserve that behaviour) or if it fails to allocate the tensor for any other reason.

In TF 1.15, a `NullPointerException` was [explicitly thrown](https://github.com/tensorflow/tensorflow/blob/8ca7de134ca196eebc43108ba79c643c2f5ac223/tensorflow/java/src/main/native/tensor_jni.cc#L358) in such case, which was a bit odd. @saudet changed it to a `IllegalStateException` in the [JavaCPP version](https://github.com/tensorflow/java/blob/1b6047804f4e6a4faf2606401502f64fc2437162/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java#L1126), which is better, but still misleading in case the mistake comes from a user-provided parameter (there is actually no way to know it from the native call).

So I've added an explicit check on the minimum tensor size requirement and now throw an `IllegalArgumentException` if that check does not pass. This is only possible with fixed-length datatypes, as we don't know the size of the elements for variable-length datatypes (`TString` factories will take care of this).